mitodl/ol-infrastructure
# Md/issue 2326

### What are the relevant tickets?
#2326 

### Description (What does it do?)
This PR implements [xqwatcher](https://github.com/mitodl/xqueue-watcher/tree/master/xqueue_watcher) + [codejail](https://github.com/openedx/codejail) following our standard practices using our PPP stack. This re-implementation completely removes the dependency on the salt-stack from xqwatcher. Additionally, it discontinues use of Ansible for installing xqwatcher. 

Key Elements / Changes:
- `src/bilder/xqwatcer`
  - Packer Config and PyInfra code for building the AMI.
  - Installs a virtual environment for xqwatcher itself
  - Installs logrotate configuration for keep xqwatcher logs under control. 
  - Installs `xqwatcher.json` and `logging.json` configuration files. 
  - Sets up (but does not configure) a new script `fetch_graders.py`. See below. 
  - Consul-templates for the following:
    - SSH key for fetching grader code from github.com / github.mit.edu 
    - `fetch_graders.py` configuration file `fetch_graders.yaml`
    - `conf.d/grader_config.json` configuration file xqwatcher
  - Additional virtual environment + configs for each course grader as needed. 
    - Includes requirements files for all grader venvs. 
    - Sets up a unix user for each grader and fixes permissions. 
    - Sets up an app-armor profile for each grader.
    - Sets up sudoers entries for each grader.
- `src/bridge/secrets/xqwatcher`
  - Contents of the files for the consul-templates outlined above. Templates are simplistic and mostly consist of rendering these secrets directly to the filesystem as-is (converting to JSON or YAML as needed). 
- `src/ol_infrastructure/xqwatcher`
  - Pretty standard Pulumi code for creating an xqwatcher stack in AWS. 
  - xqwatchers are deployed into the `operations-vpc` to facilitate connecting to more than one xqueue stack. Additional, complicated, peerings would be required if xqwatcher is deployed elsewhere. 
  - Populates values out of `src/bridge/secrets/xqwatcher` and into vault for use by consul-template.
  - Exports a security group.
- `src/ol_infrastructure/xqueue`
  - Adds the security group exported by the xqwatcher stacks to the ingress rules for xqueue.  
- Special note about `SERVER` directives in the conf.d config:
  - Because we are deploying into the `operations-vpc`, we need to further qualify the addresses of the various xqueue instances to include the consul-datacenter.
    - Previously: `http://xqueue.service.consul:8040`
    - Going forward: `http://xqueue.service.mitx-staging-qa.consul:8040`  
- `fetch_graders.py` and `fetch_graders.yaml` 
  - This script runs on startup and every hour to pull down an integrate any changes made to the grader code.  
  - The script loads the configuration yaml and clones / fetches as necessary for a given branch of a given repo and installs it into the specified directory inside of `graders`. 
  - `FETCH_GRADERS_README.md` to follow. 


### Additional Items to be done
- [x] ~`FETCH_GRADERS_README.md`~ documented in the script itself.
- [x] Build + deployment pipeline for concourse
- [ ] Non-ephemeral documentation around the `conf.d/grader_config.yaml` setup. 

### Other thoughts
- We can probably remove the graders / other elements that pretain to anything beside mit-600 and mit-686. I couldn't find any active xqwatcher instances in our environment for courses besides these two. That would clean things up a bit and shorted the build time on the AMI (potentially substantially...) 

### How can this be tested?
It is tricky. In mitx-staging-qa I have created two courses, one for 600x and one for 686x that will interact with staging's already deployed xqueue environment and submit problems. From there, you can observe the xqwatcher instance reach out to the xqueue instance, pickup a problem to grade, grade it, and return the result to xqueue, which in turn returns the result to edxapp.
